src/LCM2ROSControl.cpp

```diff
@@ -429,19 +474,94 @@ void LCM2ROSControl::update(const ros::Time& time, const ros::Duration& period)
 
           positionJointIndex++;
         }
-      
     }
 
     void LCM2ROSControl::stopping(const ros::Time& time)
     {}
 
+double LCM2ROSControl::commandEffort(const std::string& joint_name, const hardware_interface::JointHandle& joint_handle, const joint_command& command, const double dt, const Behavior& behavior) {
+  // see drc_joint_command_t.lcm for explanation of gains and
+  // force calculation.
+  double q = joint_handle.getPosition();
+  double qd = joint_handle.getVelocity();
+  double f = joint_handle.getEffort();
+  double q_desired, qd_desired, f_desired;
+
+  joint_gains gains;
+
+  switch (behavior) {
+    case Behavior::NORMAL:
+      q_desired = command.position;
+      qd_desired = command.velocity;
+      f_desired = command.effort;
+      gains = command.gains;
+      break;
+    case Behavior::POSITION_CONTROL:
+      q_desired = command.position;
+      qd_desired = 0.0;
+      f_desired = 0.0;
+      gains = behavior_gain_overrides[Behavior::POSITION_CONTROL][joint_name];
+      break;
+    case Behavior::FREEZE:
+      q_desired = latched_positions[joint_name];
+      qd_desired = 0.0;
+      gains = behavior_gain_overrides[Behavior::FREEZE][joint_name];
+      break;
+    default:
+      ROS_WARN_STREAM("Unknown behavior, treating it as FREEZE");
+      q_desired = latched_positions[joint_name];
+      qd_desired = 0.0;
+      gains = behavior_gain_overrides[Behavior::FREEZE][joint_name];
+      break;
+  }
+
+  double command_effort = 
+    gains.k_q_p * ( q_desired - q ) +
+    gains.k_q_i * ( q_desired - q ) * dt +
+    gains.k_qd_p * ( qd_desired - qd) +
+    gains.k_f_p * ( f_desired - f) +
+    gains.ff_qd * ( qd ) +
+    gains.ff_qd_d * ( qd_desired ) +
+    gains.ff_f_d * ( f_desired ) +
+    gains.ff_const;
+  return command_effort;
+}
+
+double LCM2ROSControl::currentTransitionRatio() {
+  ros::Duration time_since_transition_start = ros::Time::now() - last_transition_start_time;
+  double alpha = time_since_transition_start.toSec() / last_transition_duration.toSec();
+
+  if (!std::isnormal(alpha)) {
+    return 1.0;
+  }
+
+  return std::min(std::max(alpha, 0.0), 1.0);
+}
+
+
+void LCM2ROSControl::transitionTo(Behavior new_behavior, ros::Duration transition_duration) {
+  latchCurrentPositions();
+  last_transition_start_time = ros::Time::now();
+  last_transition_duration = transition_duration;
+  previous_behavior = current_behavior;
+  current_behavior = new_behavior;
+}
+
+void LCM2ROSControl::latchCurrentPositions() {
+  for (auto iter = positionJointHandles.begin(); iter != positionJointHandles.end(); iter++) {
+    latched_positions[iter->first] = iter->second.getPosition();
+  }
```

Comment:
It seems like this is only storing the positions of the position controlled joints in the latched_positions array. We should also record the positions of the effort controlled joints right?